### PR TITLE
Marking Address as 'undeliverable' Updates Parent Account and Matching Contact Undeliverable Fields

### PR DESCRIPTION
**Pull request by dfuller114**
# Critical Changes

# Changes

# Issues Closed

# Community Ideas Delivered

# Features Intended for Future Release

# Features for Elevate Customers

# New Metadata

- Fields
    - Undeliverable__c on Address__c object
    - Undeliverable_Address__c on Contact object
    - Undevlierable_Address__c on Account object

# Deleted Metadata

### CODE REVIEW

File: src/classes/NPSP_HouseholdAccount.cls [hunk 1]
```diff
@@ -82,6 +82,10 @@ public inherited sharing class NPSP_HouseholdAccount {
                 null);
     }
 
+    public void isBillingAddressUndeliverable(Boolean isUndeliverable) {
```
Comment (lparrott) [inline src/classes/NPSP_HouseholdAccount.cls:85]: This method name is confusing to me, I expect "is..." methods to return a Boolean, not take one as a param.

File: src/classes/Addresses.cls [hunk 2]
```diff
@@ -981,11 +997,16 @@ public inherited sharing class Addresses extends fflib_SObjects {
     */
     private void updateAddrOverrides(Map<Id, Address__c> mapAddrIdAddr,
             TDTM_Runnable.DmlWrapper dmlWrapper) {
+        System.debug(mapAddrIdAddr);
```
Comment (lparrott) [inline src/classes/Addresses.cls:1000]: Debugs to remove!

File: src/classes/Addresses.cls [hunk 3]
```diff
@@ -302,16 +300,20 @@ public inherited sharing class Addresses extends fflib_SObjects {
                     NPSP_Contact npspContact = new NPSP_Contact(con);
                     // detect that the contact's current address was deleted.
                     // and if so clear any override flag so it will get the default address.
+                    if (npspContact.hasAddressOverride() == false) {
```
Comment (lparrott) [inline src/classes/Addresses.cls:303]: Does it make sense to check this after the `npspContact.currentAddress() == null` check below it, which would change the override value?

File: src/classes/Addresses.cls [hunk 4]
```diff
@@ -273,14 +270,15 @@ public inherited sharing class Addresses extends fflib_SObjects {
...
             NPSP_Address npspAddress = new NPSP_Address(addr);
+
+            NPSP_HouseholdAccount npspHouseholdAccount = new NPSP_HouseholdAccount(accountsById.get(accId));
+            npspHouseholdAccount.isBillingAddressUndeliverable(npspAddress.isUndeliverable());
+
             if (npspHouseholdAccount.billingAddressIsDifferentFrom(npspAddress)) {
                 npspHouseholdAccount.billingAddress(npspAddress);
-                listAccUpdate.add(npspHouseholdAccount.getRecord());
             }
+            listAccUpdate.add(npspHouseholdAccount.getRecord());
```
Comment (lparrott) [inline src/classes/Addresses.cls:281]: Is there really no condition to check for this now? It seems like a lot more Accounts will get updated outside of that `if` block

**Review by lparrott — COMMENTED** [review]
Great work here, I really like the Test driven approach! Just had a few comments, will look through again when it's out of Draft.

File: src/classes/NPSP_HouseholdAccount.cls (hunk 1 above)
Comment (dfuller114) [reply src/classes/NPSP_HouseholdAccount.cls:85]: updated!

File: src/classes/Addresses.cls (hunk 2 above)
Comment (dfuller114) [reply src/classes/Addresses.cls]: updated!

File: src/classes/Addresses.cls (hunk 3 above)
Comment (dfuller114) [reply src/classes/Addresses.cls:303]: yeah good call. updated!

File: src/classes/Addresses.cls (hunk 4 above)
Comment (dfuller114) [reply src/classes/Addresses.cls]: Yeah I had that on the to-do list to make that filter more specific. It's updated now.

File: src/classes/ADDR_Addresses_TEST.cls [hunk 5]
```diff
@@ -517,7 +517,304 @@ public with sharing class ADDR_Addresses_TEST {
...
+            WHERE Id IN :accountIds]);
+
+        for (Address__c address: testAddresses) {
+            Account updatedAccount = updatedAccountsById.get(address.Household_Account__c);
+            System.assertEquals(address.Undeliverable__c,
+                    updatedAccount.Undeliverable_Address__c, 'account address' +
+                            ' undeliverable status should not match the address undeliverable status when the address' +
+                            ' is not marked as the default address.');
+            for (Contact contact : updatedAccount.Contacts) {
+                System.assert(contact.is_Address_Override__c == false, 'The contact should not be marked with an ' +
```
Comment (npsp-reedestockton) [inline src/classes/ADDR_Addresses_TEST.cls:697]: The message associated with this assertion seems more like a code comment than something appropriate for a message assertion.  I'd recommend just stating what the assertion expects and if you feel the need for a comment to explain the reason for the test, do that separately.

File: src/classes/ADDR_Addresses_TEST.cls [hunk 6]
```diff
@@ -517,7 +517,304 @@ public with sharing class ADDR_Addresses_TEST {
...
+
+        Contact updatedContact = [SELECT Undeliverable_Address__c, is_Address_Override__c
+                                  FROM Contact
+                                  WHERE Id = :contact.Id];
+        System.assert(updatedContact.is_Address_Override__c == true);
+        System.assertEquals(overrideAddress.Undeliverable__c, updatedContact.Undeliverable_Address__c);
+    }
+
+    @isTest
+    private static void updatedUndeliverableDefaultAddressShouldSyncToContactOverrideNonHousehold() {
```
Comment (npsp-reedestockton) [inline src/classes/ADDR_Addresses_TEST.cls:750]: I'd recommend naming the method for what you are testing - updatedUndeliverableDefaultAddressShouldSyncToContactOverrideOrganizationAccount??

File: src/classes/ADDR_Addresses_TEST.cls [hunk 7]
```diff
@@ -517,7 +517,304 @@ public with sharing class ADDR_Addresses_TEST {
...
+
+        Test.startTest();
+            overrideAddress.Undeliverable__c = true;
+            update overrideAddress;
+        Test.stopTest();
+
+        Contact updatedContact = [SELECT Undeliverable_Address__c, is_Address_Override__c
+                                  FROM Contact
+                                  WHERE Id = :contact.Id];
+        System.assert(updatedContact.is_Address_Override__c == true);
```
Comment (npsp-reedestockton) [inline src/classes/ADDR_Addresses_TEST.cls:745]: are you omitting the messages intentionally??  here and in a few other spots...  I don't necessarily have an issue with it, but I'm unclear on why sometimes yes and somethings no....

**Review by npsp-reedestockton — APPROVED** [review]
I made a couple of comments.  You may want to change some things based on those comments, but they are low priority, judgement calls having to do with names and messages, so I am approving.  I'll leave the changes to your discretion.

Really nice job @dfuller114 !!!

**Review by lparrott — APPROVED** [review]
Updates look good!

**Review by bdvorachek — APPROVED** [review]
All of the renaming in ADDR_Addresses_TEST and other test classes is :100:, made this more reviewable. Re-started 2GP Feature test flow, but I think this is good once that build passes!

File: src/classes/ADDR_Addresses_TEST.cls (hunk 5 above)
Comment (dfuller114) [reply src/classes/ADDR_Addresses_TEST.cls:697]: updated!

**Pull request merged**